milli/src/search/distinct/mod.rs

```diff
@@ -0,0 +1,21 @@
+mod facet_distinct;
+mod map_distinct;
+mod noop_distinct;
+
+use roaring::RoaringBitmap;
+
+pub use facet_distinct::FacetDistinct;
+pub use map_distinct::MapDistinct;
+pub use noop_distinct::NoopDistinct;
+use crate::DocumentId;
+
+pub trait DocIter: Iterator<Item=anyhow::Result<DocumentId>> {
+    /// Returns ownership on the internal RoaringBitmaps: (candidates, excluded)
+    fn into_excluded(self) -> RoaringBitmap;
+}
+
+pub trait Distinct<'a> {
+    type Iter: DocIter;
+
+    fn distinct(&'a mut self, candidates: RoaringBitmap, excluded: RoaringBitmap) -> Self::Iter;
```

Comment:
yes, because implementation of `MapDistinctIter` holds a ref to `MapDistinct`, but that lifetime can't be 'a since it would require that the ref to the map lives as long as the txn. I need a lifetime to say that this reference is gonna live at most as long as self, and this is the only way I have managed to do it without GAT.

```rust
pub struct MapDistinctIter<'a, 'b> {
    distinct: FieldId,
    map: &'b mut HashMap<String, usize>,
    index: &'a Index,
    txn: &'a heed::RoTxn<'a>,
    candidates: roaring::bitmap::IntoIter,
    excluded: RoaringBitmap,
}
```